### PR TITLE
Add api metrics for MS

### PR DESCRIPTION
**Pull request by ostolero**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes:[ link to the issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-4316)

## Description
Add API metrics for MS 

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/MicrosoftDefenderAdvancedThreatProtection/ReleaseNotes/1_13_2.md [hunk 1]
```diff
@@ -1,7 +1,8 @@
+
```
Comment (JudahSchwartz) [inline Packs/MicrosoftDefenderAdvancedThreatProtection/ReleaseNotes/1_13_2.md:1]:
```suggestion
```

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py [hunk 2]
```diff
@@ -524,6 +526,32 @@ def run_retry_on_rate_limit(self, args_for_next_run: dict):
...
+            raise DemistoException(err_msg, res=res)
+        except ValueError:
+            err_msg += '\n{}'.format(res.text)
+            raise DemistoException(err_msg, res=res)
+
+    def create_api_metrics(self, status_code):
+        execution_metrics = ExecutionMetrics()
+        ok_codes = (200, 201, 202, 204, 206)
+
+        if not execution_metrics.is_supported() or demisto.command() == 'test-module':
```
Comment (JudahSchwartz) [inline Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py:545]: `or demisto.command() == 'test-module':`
is this neccessary? Shouldnt we not have gotten here due to overwrite_rate_limit_retry? Better to keep the logic of whether or not we're running this flow in 1 place

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py [hunk 3]
```diff
@@ -524,6 +526,32 @@ def run_retry_on_rate_limit(self, args_for_next_run: dict):
                               scheduled_command=ScheduledCommand(command=demisto.command(), next_run_in_seconds=60,
                                                                  args=args_for_next_run))
 
+    def handle_error_with_metrics(self, res):
+        self.create_api_metrics(res.status_code)
+        err_msg = 'Error in API call [{}] - {}' .format(res.status_code, res.reason)
```
Comment (JudahSchwartz) [inline Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py:531]: looks like this is a copy of what we have from CSP error handling. Ideally I would say to refactor that into a method and wrap it here
If we cant then lets a comment here of where this logic came from

**Review by JudahSchwartz — COMMENTED** [review]
Amazing work on this one!
Please add screenshots of what the added widgets and the updated widget look like
Great work!

File: Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py (hunk 2 above)
Comment (ostolero) [reply Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py:545]: We didn't skip the test module in the retry mechanism, but you're right it makes more sense to skip them now. I'll add the overwrite_rate_limit_retry to them :)

File: Packs/Base/ReleaseNotes/1_31_31.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+#### Integrations
+##### O365 Outlook Mail (Using Graph API)
+- Added support for API metrics.
```
Comment (JudahSchwartz) [inline Packs/Base/ReleaseNotes/1_31_31.md:3]: we didnt add this to Base afaik

**Review by JudahSchwartz — APPROVED** [review]

File: Packs/Base/ReleaseNotes/1_31_31.md (hunk 4 above)
Comment (ostolero) [reply Packs/Base/ReleaseNotes/1_31_31.md:3]: reverted

File: Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.yml [hunk 5]
```diff
@@ -4641,12 +4641,6 @@ script:
       isArray: false
       name: ran_once_flag
       required: false
-    - default: true
-      description: ''
-      isArray: true
-      name: endpoint
```
[the remark this answers is not in the document]
Comment (ostolero) [reply Packs/MicrosoftDefenderAdvancedThreatProtection/Integrations/MicrosoftDefenderAdvancedThreatProtection/MicrosoftDefenderAdvancedThreatProtection.yml]: I just reverted this, I think it was added by format somehow. It wasn't in the original file

**Pull request merged**